crates/typst/src/doc.rs

```diff
@@ -42,17 +42,48 @@ pub struct Frame {
     baseline: Option<Abs>,
     /// The items composing this layout.
     items: Arc<Vec<(Point, FrameItem)>>,
+    /// The hardness of this frame.
+    kind: FrameKind,
 }
 
 /// Constructor, accessors and setters.
 impl Frame {
-    /// Create a new, empty frame.
+    /// Create a new, empty soft frame.
     ///
     /// Panics the size is not finite.
     #[track_caller]
-    pub fn new(size: Size) -> Self {
+    pub fn soft(size: Size) -> Self {
         assert!(size.is_finite());
-        Self { size, baseline: None, items: Arc::new(vec![]) }
+        Self {
+            size,
+            baseline: None,
+            items: Arc::new(vec![]),
+            kind: FrameKind::Soft,
+        }
+    }
+
+    /// Create a new, empty hard frame.
+    ///
+    /// Panics the size is not finite.
+    #[track_caller]
+    pub fn hard(size: Size) -> Self {
+        assert!(size.is_finite());
+        Self {
+            size,
+            baseline: None,
+            items: Arc::new(vec![]),
+            kind: FrameKind::Hard,
+        }
+    }
+
+    /// Sets the frame's hardness
```

Comment:
Based on previous style recommendations from Laurenz
```suggestion
    /// Sets the frame's hardness.
```